webapp/components/post_view/components/post_list.jsx

```diff
@@ -50,12 +54,96 @@ export default class PostList extends React.Component {
             this.introText = this.getArchivesIntroMessage();
         }
 
+        let lastViewed = Number.MAX_VALUE;
+        const member = ChannelStore.getMember(props.channel.id);
+        if (member != null) {
+            lastViewed = member.last_viewed_at;
+        }
+
         this.state = {
             isScrolling: false,
-            topPostId: null
+            topPostId: null,
+            lastViewed,
+            ownNewMessage: false
         };
     }
 
+    componentWillReceiveProps(nextProps) {
+        if (this.props.channel !== nextProps.channel) {
+            let lastViewed = Number.MAX_VALUE;
+            const member = ChannelStore.getMember(nextProps.channel.id);
+            if (member != null) {
+                lastViewed = member.last_viewed_at;
+            }
+            this.setState({
+                lastViewed,
+                ownNewMessage: false
+            });
+        }
+    }
+
+    setUnreadPost(post) {
```

Comment:
By "setting of data" I mean lines like:
```
AsyncClient.updateLastViewedAt();
ChannelStore.resetCounts(ChannelStore.getCurrentId());
ChannelStore.emitChange();
```
That's what we would call an action and we should move these lines to a reusable place (aka inside `post_actions.jsx`). Basically React components should not be modifying the stores directly, we should have a unidirectional data flow (simplified: stores -> components -> actions -> stores)